### PR TITLE
[4.0]  Add prepared statements for mod_related_items

### PR DESCRIPTION
**Pull request by HLeithner**
### Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the module in all ways you can think of.

* create a frontend module
* create articles with the same metatags
* You should see related items based on the metatags matching

### Expected result
Nothing changed.

### CODE REVIEW

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 1]
```diff
@@ -104,30 +107,48 @@ public static function getList(&$params)
...
+				$binds  = [];
+				$wheres = [];
 
 				foreach ($likes as $keyword)
 				{
-					$wheres[] = 'a.metakey LIKE ' . $db->quote('%' . $keyword . '%');
+					$binds[] = '%' . $keyword . '%';
+				}
+
+				$bindNames = $query->bindArray($binds, parameterType::STRING);
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]:
```suggestion
				$bindNames = $query->bindArray($binds, ParameterType::STRING);
```

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 2]
```diff
@@ -71,7 +73,8 @@ public static function getList(&$params)
 			// Select the meta keywords from the item
 			$query->select('metakey')
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]: Apply `$db->quoteName`

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 3]
```diff
@@ -104,30 +107,48 @@ public static function getList(&$params)
...
+
+				foreach ($bindNames as $keyword)
+				{
+					$wheres[] = 'a.metakey LIKE ' . $keyword;
 				}
 
 				$query->where('(' . implode(' OR ', $wheres) . ')')
-					->where('(a.publish_up = ' . $db->quote($nullDate) . ' OR a.publish_up <= ' . $db->quote($now) . ')')
-					->where('(a.publish_down = ' . $db->quote($nullDate) . ' OR a.publish_down >= ' . $db->quote($now) . ')');
+					->where('(a.publish_up = :nullDate1 OR a.publish_up <= :nowDate1)')
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]: Apply `$db->quoteName`

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 4]
```diff
@@ -35,12 +36,13 @@ abstract class RelatedItemsHelper
...
-		$groups  = implode(',', Factory::getUser()->getAuthorisedViewLevels());
-		$maximum = (int) $params->get('maximum', 5);
-		$factory = $app->bootComponent('com_content')->getMVCFactory();
+		$db        = Factory::getDbo();
+		$app       = Factory::getApplication();
+		$input     = $app->input;
+		$groups    = Factory::getUser()->getAuthorisedViewLevels();
+		$maximum   = (int) $params->get('maximum', 5);
+		$factory   = $app->bootComponent('com_content')->getMVCFactory();
+		$condition = ContentComponent::CONDITION_PUBLISHED;
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: Remove variable?

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: this will be removed when workflow move more into a state machine

Comment (Quy) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: Remove the unnecessary variable and insert it here.
`->bind(':condition', $condition, ParameterType::INTEGER);`

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: As explained this will be changed in the hopefully near future

Comment (brianteeman) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: That may be true - it may be wishful thinking. Either way all code that is committed should be correct at the time it is committed.

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php:43]: The code is not wrong.

**Review by alikon — APPROVED** [review]

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 29acfd9eccb13d682760092d982487a7593bccf1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25042">issues.joomla.org/tracker/joomla-cms/25042</a>.</sub>

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 5]
```diff
@@ -102,30 +105,48 @@ public static function getList(&$params)
...
+
+				foreach ($bindNames as $keyword)
+				{
+					$wheres[] = 'a.metakey LIKE ' . $keyword;
 				}
 
 				$query->where('(' . implode(' OR ', $wheres) . ')')
-					->where('(a.publish_up = ' . $db->quote($nullDate) . ' OR a.publish_up <= ' . $db->quote($now) . ')')
-					->where('(a.publish_down = ' . $db->quote($nullDate) . ' OR a.publish_down >= ' . $db->quote($now) . ')');
+					->where('('. $db->quoteName('a.publish_up') . ' = :nullDate1 OR ' . $db->quoteName('a.publish_up') . ' <= :nowDate1)')
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]:
```suggestion
					->where('(' . $db->quoteName('a.publish_up') . ' = :nullDate1 OR ' . $db->quoteName('a.publish_up') . ' <= :nowDate1)')
```

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 6]
```diff
@@ -102,30 +105,48 @@ public static function getList(&$params)
...
+					->where('('. $db->quoteName('a.publish_down') . ' = :nullDate2 OR ' . $db->quoteName('a.publish_down') . ' >= :nowDate2)')
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]:
```suggestion
					->where('(' . $db->quoteName('a.publish_down') . ' = :nullDate2 OR ' . $db->quoteName('a.publish_down') . ' >= :nowDate2)')
```

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 7]
```diff
@@ -102,30 +105,48 @@ public static function getList(&$params)
...
 				{
-					$wheres[] = 'a.metakey LIKE ' . $db->quote('%' . $keyword . '%');
+					$binds[] = '%' . $keyword . '%';
+				}
+
+				$bindNames = $query->bindArray($binds, ParameterType::STRING);
+
+				foreach ($bindNames as $keyword)
+				{
+					$wheres[] = 'a.metakey LIKE ' . $keyword;
```
Comment (SharkyKZ) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]:
```suggestion
					$wheres[] = $db->quoteName('a.metakey') . ' LIKE ' . $keyword;
```

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 8]
```diff
@@ -102,30 +105,48 @@ public static function getList(&$params)
...
-					->select('a.id')
-					->from('#__content AS a')
-					->where('a.id != ' . (int) $id)
-					->where('ws.condition = ' . ContentComponent::CONDITION_PUBLISHED)
-					->where('a.access IN (' . $groups . ')');
-
-				$wheres = array();
+					->select($db->quoteName('a.id'))
+					->from($db->quoteName('#__content', 'a'))
+					->leftJoin($db->quoteName('#__workflow_associations', 'wa'), $db->quoteName('wa.item_id') . ' = ' . $db->quoteName('a.id'))
```
Comment (Quy) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]: Convert to `join('LEFT',`

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 533ead7b929858f7b6395da6362a25d764625515<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25042">issues.joomla.org/tracker/joomla-cms/25042</a>.</sub>

**Comment by ghost** [comment]
@alikon can you please retest?

File: modules/mod_related_items/Helper/RelatedItemsHelper.php [hunk 9]
```diff
@@ -102,30 +105,46 @@ public static function getList(&$params)
...
+
+				foreach ($bindNames as $keyword)
+				{
+					$wheres[] = $db->quoteName('a.metakey') . ' LIKE ' . $keyword;
 				}
 
 				$query->where('(' . implode(' OR ', $wheres) . ')')
-					->where('(a.publish_up = ' . $db->quote($nullDate) . ' OR a.publish_up <= ' . $db->quote($now) . ')')
-					->where('(a.publish_down = ' . $db->quote($nullDate) . ' OR a.publish_down >= ' . $db->quote($now) . ')');
+					->where('(' . $db->quoteName('a.publish_up') . ' = :nullDate1 OR ' . $db->quoteName('a.publish_up') . ' <= :nowDate1)')
```
Comment (alikon) [inline modules/mod_related_items/Helper/RelatedItemsHelper.php]: we can use `orWhere` but this can be matter for another pr

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: I think we don't use orWhere instead we don't use the non alias function extendWhere correct?

should it be
```->extendWhere('OR', [$db->quoteName('a.publish_up') . ' = :nullDate1', $db->quoteName('a.publish_up') . ' <= :nowDate1'])```
?

Comment (alikon) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: personal taste, i would prefer `orWhere ` cause is more readable imo, but...

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: https://github.com/joomla-framework/database/issues/174#issuecomment-509511291

In this issue comment you would remove the orWhere function. now I'm confused.

Comment (alikon) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: never mentioned `orWhere`  there ,  but as it is a personal taste go with `extendWhere`

Comment (Quy) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: Do it like this: https://github.com/joomla/joomla-cms/pull/25977/files#diff-759da5f761521231cbefc3bd96b7846cR135

Comment (HLeithner) [reply modules/mod_related_items/Helper/RelatedItemsHelper.php]: Updated

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on f79b14950b01be6414323a3a8d0f6164d55e04c0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25042">issues.joomla.org/tracker/joomla-cms/25042</a>.</sub>

**Comment by ghost** [comment]
@SharkyKZ can you please retest?

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f79b14950b01be6414323a3a8d0f6164d55e04c0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25042">issues.joomla.org/tracker/joomla-cms/25042</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25042">issues.joomla.org/tracker/joomla-cms/25042</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!
